docker/bookkeeper/entrypoint.sh

```diff
@@ -122,10 +124,14 @@ initialize_cluster() {
             /opt/bookkeeper/bin/bookkeeper shell initnewcluster
             if [ $? -eq 0 ]; then
                 echo "initnewcluster operation succeeded"
+                zk-shell --run-once "rm ${BK_CLUSTER_ROOT_PATH}/bkInitLock" ${BK_zkServers}
```

Comment:
Alright, this looks better, but what happens if the process running this script crashes between creating the lock znode and deleting it? Are we going to be in a situation that the lock is held forever by a non-existing process?

Reply:
If that happens, the persistent node will be left behind in Zookeeper. 
A remedy would be to manually deleting that znode from Zookeeper cluster.

Reply:
Yeah, I don't think this is going to work well. Here is a different proposal:

1- Change it back to ephemeral
2- Run `zk-shell` without `--run-once` and in a background process. Make sure to get the PID of the background process right after creating it.
3- Once you are done with the bk initialization, kill PID.

With this proposal, if the container crashes, then all processes die and the session is eventually closed removing the ephemeral. In the absence of crashes, it removes the session after being done with the initialization. The only missing case that is not covered is the one of the parent process dying. In this case, the child process becomes orphaned and the session is never closed. If we can find a way to have the child process exiting with the parent, then this could be a nice solution.

Reply:
Ok, I think this SO thread gives a solution to the last point by using `trap`:

https://stackoverflow.com/questions/360201/how-do-i-kill-background-processes-jobs-when-my-shell-script-exits/2173421

See if this new proposal works, please.

Reply:
@fpj I think this change would be a significant one (especially w.r.t. testing negative scenarios), and we should do this as part of a separate enhancement issue. 

I can create an enhancement issue if you agree. Better still, we can try this strategy out separately in the Apache/Bookkeeper image improvement PR that is currently in progress: https://github.com/apache/bookkeeper/pull/2219.

Reply:
I don't know why that's a significant change, it is a four-line change!

1- Change it back to ephemeral -> Only changes a parameter in the existing command
2- Run zk-shell without --run-once and in a background process. Make sure to get the PID of the background process right after creating it. -> Changes the command and you have to add `&` to the end of the command. Get the PID of the background process with `$!`, which another line but only an assignment.. 
3- Once you are done with the bk initialization, kill PID. -> Replace the deletes with kill PID.
4- Add a trap call to kill the child in the case of exiting prematurely.

Have you even tried?

Reply:
I meant it is a significant change in the sense of target behavior and the impact on testing. And, yes I have been trying to make that change - not done yet.

Reply:
When trying to test out the suggestion, I found this issue. We need the output of the `zk-shell` command (the exit status is always `0`) to verify if the lock was created successfully, which cannot be done using `trap` AFAIK. Am I missing something?

https://github.com/pravega/pravega/blob/6da3c0ed1627f6be9c44a37023b8774ba2efcc4b/docker/bookkeeper/entrypoint.sh#L116

Besides, taking a step back and reconsidering the discussion here, I realize that the original code that created the ephemeral node inline was just fine. Here's why: 

1. **Scenario 1:** All nodes attempt to come up simultaneously. All see `${BK_zkLedgersRootPath}/available/readonly` znode missing (as the cluster is not initialized yet). One node successfully creates `${BK_CLUSTER_ROOT_PATH}/bkInitLock` znode and initializes the cluster, others see that `${BK_CLUSTER_ROOT_PATH}/bkInitLock`znode is already present, so they don't attempt to initialize the cluster. Business as usual, and all nodes come up successfully.

2. **Scenario 2:** One node comes up first, and creates the `${BK_CLUSTER_ROOT_PATH}/bkInitLock` znode, successfully initilizes the cluster (which in turn successfully creates `${BK_zkLedgersRootPath}/available/readonly`). Others come up later, they see `${BK_zkLedgersRootPath}/available/readonly`, so they don't attempt to initialize the cluster and run successfully. 

3. **Scenario 3:** All nodes see `${BK_zkLedgersRootPath}/available/readonly` znode missing (so the cluster is not initialized yet). All of them even find `${BK_CLUSTER_ROOT_PATH}/bkInitLock` znode missing, so each one tries to create the znode. Only one succeeds in creating the znode and it also initializes the cluster, the others get a non-empty output when attempting to create the znode, so they exit. The orchestration engine (Kubernetes/Docker Swarm) brings the instance back up and they start running successfully, as they find `${BK_zkLedgersRootPath}/available/readonly` znode is present and don't attempt to initialize the cluster this time. 

4. **Scenario 4:** All nodes see `${BK_zkLedgersRootPath}/available/readonly` znode missing (so the cluster is not initialized yet). All of them even find `${BK_CLUSTER_ROOT_PATH}/bkInitLock` znode missing, so each one tries to create the znode. Only one succeeds in creating the znode and it also initializes the cluster. By the others wake up from slumber, the ephemeral `${BK_CLUSTER_ROOT_PATH}/bkInitLock` is gone, and therefore they try to initialize the cluster. Init cluster operation fails and they exit. Like in the previous case the orchestration engine brings the instance back up and they start running successfully, as they find `${BK_zkLedgersRootPath}/available/readonly` znode is present and don't attempt to initialize the cluster this time.

Reply:
If what you say above is true, in particular for scenario 4, then we don't need a lock at all. If we do need a lock, then reverting the change will give you a broken lock implementation. The question we need to answer is whether the lock is strictly needed or not.

Reply:
Without the `${BK_CLUSTER_ROOT_PATH}/bkInitLock`, in the scenarios where all instances didn't find that the cluster was already initialized, n-1 instances will always unnecessarily attempt to initialize the cluster, wait for the operation to fail, and will come up only upon restart. By use of the lock, we reduce (but not eliminate) the chances of that happening. The downside of the current lock implementation is that it doesn't strictly prevent such situations, but it does reduce such situations. 

Nonetheless, I see merit in strengthening the lock. I suggest we do that as part of a separate issue.

Reply:
Is it the case that if two instances try to initialize concurrently, then only one will succeed? If that's the case, then mutual exclusion is already implemented, and we don't strictly need a lock. Could you please confirm that this interpretation is correct? 

If it is correct, then the "lock" is not really a lock, it is only a convenient signal to avoid unnecessary restarts, and as such, it can remain as is.

Reply:
Yes, that is right - if two instances attempt to initialize concurrently, only one will succeed. The others will exit (and will be restarted by the orchestrator). 

https://github.com/pravega/pravega/blob/6da3c0ed1627f6be9c44a37023b8774ba2efcc4b/docker/bookkeeper/entrypoint.sh#L133

Sure, it sounds more like a signal. :)

I'll revert it back to ephemeral znode as before.

Reply:
Done.

Reply:
> Yes, that is right - if two instances attempt to initialize concurrently, only one will succeed. 

What evidence do you have that this is correct and that concurrent attempts won't lead to some form of unexpected corruption due to a race?

Reply:
I don't have hard evidence that anything wrong cannot happen. What I said is based on the logic in the code, as well as casual observations made from looking at the logs from various runs during this effort of fixing the BK image.

Reply:
This is the code that initializes a new cluster:

https://github.com/apache/bookkeeper/blob/branch-4.8/bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationManager.java#L419

There are two safety guards. First, it checks whether the ledger root path exists. It could happen that two instances cross that check concurrently, though. Second, the operations that need to be executed against zookeeper are executed in a multi transaction (all or nothing). If one fails, then none is applied and the call throws, which will happen in the case of concurrent attempts.

This code is safe without the `bkInitLock` that is being discussed here.